### ISSUE TITLE
<cmd>:<num>: command not found: <expr>

### ISSUE THREAD

**Issue by xh3b4sd** [robbyrussell/oh-my-zsh#5490]
Hello. I am using `oh-my-zsh` and found an issue. In the first place I thought it was an issue in `ack-grep`, because this was the command I experienced the issue first. Then it turned out it probably is more about the shell I am using. So here you go. 

```
vagrant@xenial ack-grep --version
ack 2.14
Running under Perl 5.22.1 at /usr/bin/perl

Copyright 2005-2014 Andy Lester.

This program is free software.  You may modify or distribute it
under the terms of the Artistic License v2.0.
vagrant@xenial ack-grep Foo().Bar
vagrant@xenial ack-grep
ack-grep:2: command not found: .Bar
vagrant@xenial ack-grep --version
ack-grep:3: command not found: .Bar
```
```
vagrant@xenial zsh --version
zsh 5.1.1 (x86_64-ubuntu-linux-gnu)
```
```
vagrant@xenial tar --version
tar (GNU tar) 1.28
Copyright (C) 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by John Gilmore and Jay Fenlason.
vagrant@xenial tar Foo().Bar
vagrant@xenial tar --version
tar:3: command not found: .Bar
```

See also the original issue in the `ack-grep` repo: https://github.com/petdance/ack2/issues/612. Have somebody an idea what is going on here?

**Comment by mcornella** [comment]
That's actually a zsh syntax error running without errors. Your statement turns `tar` into a function. See:

```zsh
$ tar Foo().Bar
$ which tar
tar () {
	.Bar
}
```

Now, I don't know if that's intended behavior or it should throw a syntax error (which you get on bash), but in any case this would never work. Parenthesis, curly braces and square brackets are symbols that are interpreted by the shell, so you have to quote them always: `tar 'Foo().Bar'`. That and certain globbing characters like asterisk (`*`), question mark (`?`) and equal sign (`=`).

TL;DR: always be quoting.

**Comment by xh3b4sd** [comment]
IMO this issue points to a problem that is not solved. Closing this issue is thus not the right thing to do. I would rather appreciate trying to fix the problem instead of telling the user to fix is behaviour, which never works. Software needs to be fixed, not the user. Since the technical details of your explanation are correct, the user does not care at all about them. The user has a problem, which still exists, even though you are closing the issue.

**Comment by mcornella** [comment]
Ok maybe I'm not explaining myself correctly. Zsh and Bash and other shells use parenthesis to define functions. There's no going around that. If you use parenthesis you have to tell to the shell that you mean a string which contains parenthesis, which you do by quoting the argument. There is no *software to be fixed*.

You define functions with parenthesis. You de define strings that contain parenthesis by putting them inside quotes. That's just the shell language syntax. And so is in many other languages. Just quote the string and it'll work.

Now, you can make the argument that the error it spits out is confusing. The way bash works is spitting a syntax error:

```
$ tar Foo().Bar
bash: syntax error near unexpected token `('
```

So probably zsh should just output a syntax error when doing that. I'm not exactly sure why that doesn't happen, but you'll need to ask that over to the [official zsh mailing list](http://www.zsh.org/mla/). Hope that clears things up.

**Comment by xh3b4sd** [comment]
I see this might not an issue related to `oh-my-zsh`. Closing this issue here is fine then. What is not correct thought is `There is no software to be fixed`. Anyway, not your problem. Thanks clarifying the situation.

**Comment by mcornella** [comment]
I mean that you'll have the same problem when trying to use it with ack-grep, only that there would be a syntax error.

But maybe that is a valid function definition statement and I don't know about it. Anyway the folks at zsh will know about it better than I do.
